controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -218,14 +245,51 @@ func (w *Workload) updateCoreDNSDeployment(ctx context.Context, info *coreDNSInf
 	return helper.Patch(ctx, info.Deployment)
 }
 
-// UpdateCoreDNSImageInfoInKubeadmConfigMap updates the kubernetes version in the kubeadm config map.
+// updateCoreDNSImageInfoInKubeadmConfigMap updates the kubernetes version in the kubeadm config map.
 func (w *Workload) updateCoreDNSImageInfoInKubeadmConfigMap(ctx context.Context, dns *bootstrapv1.DNS, version semver.Version) error {
 	return w.updateClusterConfiguration(ctx, func(c *bootstrapv1.ClusterConfiguration) {
 		c.DNS.ImageRepository = dns.ImageRepository
 		c.DNS.ImageTag = dns.ImageTag
 	}, version)
 }
 
+// updateCoreDNSClusterRole updates the CoreDNS ClusterRole when necessary.
+// CoreDNS >= 1.8.1 uses EndpointSlices. kubeadm < 1.22 doesn't include the EndpointSlice rule in the CoreDNS ClusterRole.
+// To support Kubernetes clusters >= 1.22 (which have been initialized with kubeadm < 1.22) with CoreDNS versions >= 1.8.1
+// we have to update the ClusterRole accordingly.
+func (w *Workload) updateCoreDNSClusterRole(ctx context.Context, kubernetesVersion semver.Version, info *coreDNSInfo) error {
+	// Do nothing for Kubernetes < 1.22.
+	if kubernetesVersion.LT(semver.Version{Major: 1, Minor: 22, Patch: 0}) {
+		return nil
+	}
+
+	// Do nothing for CoreDNS < 1.8.1.
+	targetCoreDNSVersion, err := extractImageVersion(info.ToImageTag)
+	if err != nil {
+		return err
+	}
+	if targetCoreDNSVersion.LT(semver.Version{Major: 1, Minor: 8, Patch: 1}) {
+		return nil
+	}
+
+	key := ctrlclient.ObjectKey{Name: coreDNSClusterRoleName, Namespace: metav1.NamespaceSystem}
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		var currentClusterRole rbacv1.ClusterRole
+		if err = w.Client.Get(ctx, key, &currentClusterRole); err != nil {
+			return fmt.Errorf("failed to get ClusterRole %q", coreDNSClusterRoleName)
+		}
+
+		if !reflect.DeepEqual(currentClusterRole.Rules, coreDNS181PolicyRules) {
+			updatedClusterRole := currentClusterRole.DeepCopy()
```

Comment:
Do we need to copy here?

Reply:
I think I was confusing it with something along the lines of "we have to deep copy otherwise we will change the object in the cache which could lead to strange bugs". But I assume that's not an issue here as we're using our own object.